File: cpp/testbls.cpp
/*
   Copyright (C) 2019 MIRACL UK Ltd.

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU Affero General Public License as
    published by the Free Software Foundation, either version 3 of the
    License, or (at your option) any later version.


    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Affero General Public License for more details.

     https://www.gnu.org/licenses/agpl-3.0.en.html

    You should have received a copy of the GNU Affero General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.

   You can be released from the requirements of the license by purchasing
   a commercial license. Buying such a license is mandatory as soon as you
   develop commercial activities involving the MIRACL Core Crypto SDK
   without disclosing the source code of your own applications, or shipping
   the MIRACL Core Crypto SDK with a closed source product.
*/

/* test driver and function exerciser for BLS Signature API Functions */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include "bls_BN254.h"

#if CHUNK==32 || CHUNK==64
#include "bls_BLS12383.h"
#include "bls192_BLS24479.h"
#include "bls256_BLS48556.h"
#endif

#include "randapi.h"

/* To reverse the groups G1 and G2, edit BLS*.cpp

Swap G1 <-> G2
Swap ECP <-> ECPn
Disable G2 precomputation
Switch G1/G2 parameter order in pairing function calls

#define REVERSE here

See BLSREV*.cpp

*/

//#define REVERSE

using namespace core;

static char message[] = "This is a test message";

int bls_BN254(csprng *RNG)
{
    using namespace BN254;

    int i,res;
    char s[BGS_BN254];
    char ikm[64];
#ifdef REVERSE
    char w[BFS_BN254+1], sig[4 * BFS_BN254 + 1];        // w is 2* if not compressed else 1*. sig is 4* if not compressed, else 2*
#else
    char w[4 * BFS_BN254 + 1], sig[BFS_BN254 + 1];      // w is 4* if not compressed else 2*. sig is 2* if not compressed, else 1*
#endif
    octet S = {0, sizeof(s), s};
    octet W = {0, sizeof(w), w};
    octet SIG = {0, sizeof(sig), sig};
    octet IKM = {0, sizeof(ikm), ikm};
    octet M = {0,sizeof(message), message};

    OCT_jstring(&M,message);

    res = BLS_INIT();
    if (res == BLS_FAIL)
    {
        printf("Failed to initialize\n");
        return res;
    }

    OCT_rand(&IKM,RNG,32);
    //IKM.len=32;
    //for (i=0;i<IKM.len;i++)
    //        IKM.val[i]=i+1;

    res = BLS_KEY_PAIR_GENERATE(&IKM, &S, &W);
    if (res == BLS_FAIL)
    {
        printf("Failed to generate keys\n");
        return res;
    }

    printf("Private key= 0x");
    OCT_output(&S);
    printf("Public key= 0x");
    OCT_output(&W);

    BLS_CORE_SIGN(&SIG, &M, &S);
    printf("Signature= 0x");
    OCT_output(&SIG);

    //message[7]='f'; // change the message

    res = BLS_CORE_VERIFY(&SIG, &M, &W);
    if (res == BLS_OK) printf("Signature is OK\n");
    else printf("Signature is *NOT* OK\n");
    return res;
}

#if CHUNK==32 || CHUNK==64

int bls_BLS12383(csprng *RNG)
{
    using namespace BLS12383;

    int i,res;
    char s[BGS_BLS12383];
    char ikm[64];
#ifdef REVERSE
    char w[BFS_BLS12383+1], sig[4 * BFS_BLS12383 + 1];
#else
    char w[4 * BFS_BLS12383 + 1], sig[BFS_BLS12383 + 1];
#endif
    octet S = {0, sizeof(s), s};
    octet W = {0, sizeof(w), w};
    octet SIG = {0, sizeof(sig), sig};
    octet IKM = {0, sizeof(ikm), ikm};
    octet M = {0,sizeof(message), message};

    OCT_jstring(&M,message);

    res = BLS_INIT();
    if (res == BLS_FAIL)
    {
        printf("Failed to initialize\n");
        return res;
    }

    OCT_rand(&IKM,RNG,32);
    //IKM.len=32;
    //for (i=0;i<IKM.len;i++)
    //        IKM.val[i]=i+1;

    res = BLS_KEY_PAIR_GENERATE(&IKM, &S, &W);
    if (res == BLS_FAIL)
    {
        printf("Failed to generate keys\n");
        return res;
    }

    printf("Private key= 0x");
    OCT_output(&S);
    printf("Public key= 0x");
    OCT_output(&W);

    BLS_CORE_SIGN(&SIG, &M, &S);
    printf("Signature= 0x");
    OCT_output(&SIG);

    //message[7]='f'; // change the message

    res = BLS_CORE_VERIFY(&SIG, &M, &W);
    if (res == BLS_OK) printf("Signature is OK\n");
    else printf("Signature is *NOT* OK\n");
    return res;
}

int bls_BLS24479(csprng *RNG)
{
    using namespace BLS24479;

    int i,res;
    char s[BGS_BLS24479];
    char ikm[64];
#ifdef REVERSE
    char w[BFS_BLS24479+1], sig[8 * BFS_BLS24479 + 1];
#else
    char w[8 * BFS_BLS24479 + 1], sig[BFS_BLS24479 + 1];
#endif
    octet S = {0, sizeof(s), s};
    octet W = {0, sizeof(w), w};
    octet SIG = {0, sizeof(sig), sig};
    octet IKM = {0, sizeof(ikm), ikm};
    octet M = {0,sizeof(message), message};

    OCT_jstring(&M,message);

    res = BLS_INIT();
    if (res == BLS_FAIL)
    {
        printf("Failed to initialize\n");
        return res;
    }

    OCT_rand(&IKM,RNG,48);
    //IKM.len=48;
    //for (i=0;i<IKM.len;i++)
    //        IKM.val[i]=i+1;

    res = BLS_KEY_PAIR_GENERATE(&IKM, &S, &W);
    if (res == BLS_FAIL)
    {
        printf("Failed to generate keys\n");
        return res;
    }

    printf("Private key= 0x");
    OCT_output(&S);
    printf("Public key= 0x");
    OCT_output(&W);

    BLS_CORE_SIGN(&SIG, &M, &S);
    printf("Signature= 0x");
    OCT_output(&SIG);

    //message[7]='f'; // change the message

    res = BLS_CORE_VERIFY(&SIG, &M, &W);
    if (res == BLS_OK) printf("Signature is OK\n");
    else printf("Signature is *NOT* OK\n");
    return res;
}

int bls_BLS48556(csprng *RNG)
{
    using namespace BLS48556;

    int i,res;
    char s[BGS_BLS48556];
    char ikm[64];
#ifdef REVERSE
    char w[BFS_BLS48556+1], sig[16 * BFS_BLS48556 + 1];
#else
    char w[16 * BFS_BLS48556 + 1], sig[BFS_BLS48556 + 1];
#endif
    octet S = {0, sizeof(s), s};
    octet W = {0, sizeof(w), w};
    octet SIG = {0, sizeof(sig), sig};
    octet IKM = {0, sizeof(ikm), ikm};
    octet M = {0,sizeof(message), message};

    OCT_jstring(&M,message);

    res = BLS_INIT();
    if (res == BLS_FAIL)
    {
        printf("Failed to initialize\n");
        return res;
    }

    OCT_rand(&IKM,RNG,64);
    //IKM.len=64;
    //for (i=0;i<IKM.len;i++)
    //        IKM.val[i]=i+1;

    res = BLS_KEY_PAIR_GENERATE(&IKM, &S, &W);
    if (res == BLS_FAIL)
    {
        printf("Failed to generate keys\n");
        return res;
    }

    printf("Private key= 0x");
    OCT_output(&S);
    printf("Public key= 0x");
    OCT_output(&W);

    BLS_CORE_SIGN(&SIG, &M, &S);
    printf("Signature= 0x");
    OCT_output(&SIG);

    //message[7]='f'; // change the message

    res = BLS_CORE_VERIFY(&SIG, &M, &W);
    if (res == BLS_OK) printf("Signature is OK\n");
    else printf("Signature is *NOT* OK\n");
    return res;
}

#endif

int main()
{
    int i, res;
    unsigned long ran;

    char raw[100];
    octet RAW = {0, sizeof(raw), raw};
    csprng RNG;                // Crypto Strong RNG

    time((time_t *)&ran);

    RAW.len = 100;              // fake random seed source
    RAW.val[0] = ran;
    RAW.val[1] = ran >> 8;
    RAW.val[2] = ran >> 16;
    RAW.val[3] = ran >> 24;
    for (i = 4; i < 100; i++) RAW.val[i] = i;

    CREATE_CSPRNG(&RNG, &RAW);  // initialise strong RNG

    printf("%d bit build\n", CHUNK);

    printf("\nTesting BLS signature for curve BN254\n");
    bls_BN254(&RNG);

#if CHUNK!=16
    printf("\nTesting BLS signature for curve BLS383\n");
    bls_BLS12383(&RNG);

    printf("\nTesting BLS signature for curve BLS24479\n");
    bls_BLS24479(&RNG);

    printf("\nTesting BLS signature for curve BLS48556\n");
    bls_BLS48556(&RNG);
#endif

    KILL_CSPRNG(&RNG);
}

File: swift/TestBLS.swift
/*
 * Copyright (c) 2012-2020 MIRACL UK Ltd.
 *
 * This file is part of MIRACL Core
 * (see https://github.com/miracl/core).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* Test driver and function exerciser for Boneh-Lynn-Shacham BLS Signature API Functions */

/* To reverse the groups G1 and G2, edit BLS*.go

Swap G1 <-> G2
Swap ECP <-> ECPn
Disable G2 precomputation
Switch G1/G2 parameter order in pairing function calls

Swap G1S and G2S in this program

*/

//
//  TestBLS.swift
//
//  Created by Michael Scott on 02/07/2015.
//  Copyright (c) 2015 Michael Scott. All rights reserved.
//

/* Test Boneh-Lynn-Shacham signature API */

import Foundation
import core // comment out for Xcode
import bn254
import bls12383
import bls24479
import bls48556

public func printBinary(_ array: [UInt8])
{
    for i in 0 ..< array.count
    {
        let h=String(format:"%02x",array[i])
        print("\(h)", terminator: "")
    }
    print(" ");
}


public func TestBLS_bn254(_ rng: inout RAND)
{
    let BGS=bn254.BLS.BGS
    let BFS=bn254.BLS.BFS
    let G1S=BFS+1    // Group 1 Size - compressed
    let G2S=2*BFS+1;     // Group 2 Size - compressed
   
    var S=[UInt8](repeating: 0,count: BGS)
    var W=[UInt8](repeating: 0,count: G2S)
    var SIG=[UInt8](repeating: 0,count: G1S)
    var IKM=[UInt8](repeating: 0,count: 32)

    for i in 0..<32 {
        //IKM[i]=UInt8(i+1)
        IKM[i]=rng.getByte()
    }

    var res=bn254.BLS.initit()
    if res != 0 {
        print("Failed to Initialize\n")
        return
    }

    res=bn254.BLS.KeyPairGenerate(IKM,&S,&W)
    if res != 0 {
        print("Failed to Generate Keys\n")
        return
    }

    print("Testing bn254 curve\n")
    print("Private key : 0x",terminator: "");  printBinary(S);
    print("Public  key : 0x",terminator: "");  printBinary(W);
    let mess="This is a test message"
    print("Signing string - "+mess)

    bn254.BLS.core_sign(&SIG,[UInt8](mess.utf8),S)
    print("Signature : 0x",terminator: "");  printBinary(SIG);

    res=bn254.BLS.core_verify(SIG,[UInt8](mess.utf8),W)
    if res == 0 {
        print("Signature is OK\n");
    } else {
        print("Signature is *NOT* OK\n");
    }

}

public func TestBLS_bls12383(_ rng: inout RAND)
{
    let BGS=bls12383.BLS.BGS
    let BFS=bls12383.BLS.BFS
    let G1S=BFS+1    // Group 1 Size - compressed
    let G2S=2*BFS+1;     // Group 2 Size - compressed
   
    var S=[UInt8](repeating: 0,count: BGS)
    var W=[UInt8](repeating: 0,count: G2S)
    var SIG=[UInt8](repeating: 0,count: G1S)
    var IKM=[UInt8](repeating: 0,count: 32)

    for i in 0..<32 {
        //IKM[i]=UInt8(i+1)
        IKM[i]=rng.getByte()
    }

    var res=bls12383.BLS.initit()
    if res != 0 {
        print("Failed to Initialize\n")
        return
    }
    res=bls12383.BLS.KeyPairGenerate(IKM,&S,&W)
    if res != 0 {
        print("Failed to Generate Keys\n")
        return
    }
    print("Testing bls12383 curve\n")    
    print("Private key : 0x",terminator: "");  printBinary(S);
    print("Public  key : 0x",terminator: "");  printBinary(W);
    let mess="This is a test message"
    print("Signing string - "+mess)    

    bls12383.BLS.core_sign(&SIG,[UInt8](mess.utf8),S)
    print("Signature : 0x",terminator: "");  printBinary(SIG);

    res=bls12383.BLS.core_verify(SIG,[UInt8](mess.utf8),W)
    if res == 0 {
        print("Signature is OK\n");
    } else {
        print("Signature is *NOT* OK\n");
    }

}

public func TestBLS_bls24479(_ rng: inout RAND)
{
    let BGS=bls24479.BLS192.BGS
    let BFS=bls24479.BLS192.BFS
    let G1S=BFS+1    // Group 1 Size  - compressed
    let G2S=4*BFS+1;     // Group 2 Size - compressed
   
    var S=[UInt8](repeating: 0,count: BGS)
    var W=[UInt8](repeating: 0,count: G2S)
    var SIG=[UInt8](repeating: 0,count: G1S)
    var IKM=[UInt8](repeating: 0,count: 48)

    for i in 0..<48 {
        //IKM[i]=UInt8(i+1)
        IKM[i]=rng.getByte()
    }

    var res=bls24479.BLS192.initit()
    if res != 0 {
        print("Failed to Initialize\n")
        return
    }
    res=bls24479.BLS192.KeyPairGenerate(IKM,&S,&W)
    if res != 0 {
        print("Failed to Generate Keys\n")
        return
    }
    print("Testing bls24479 curve\n")      
    print("Private key : 0x",terminator: "");  printBinary(S);
    print("Public  key : 0x",terminator: "");  printBinary(W);
    let mess="This is a test message"
    print("Signing string - "+mess)

    bls24479.BLS192.core_sign(&SIG,[UInt8](mess.utf8),S)
    print("Signature : 0x",terminator: "");  printBinary(SIG);

    res=bls24479.BLS192.core_verify(SIG,[UInt8](mess.utf8),W)
    if res == 0 {
        print("Signature is OK\n");
    } else {
        print("Signature is *NOT* OK\n");
    }
}

public func TestBLS_bls48556(_ rng: inout RAND)
{
    let BGS=bls48556.BLS256.BGS
    let BFS=bls48556.BLS256.BFS
    let G1S=BFS+1    // Group 1 Size - compressed
    let G2S=8*BFS+1;     // Group 2 Size - compressed
   
    var S=[UInt8](repeating: 0,count: BGS)
    var W=[UInt8](repeating: 0,count: G2S)
    var SIG=[UInt8](repeating: 0,count: G1S)
    var IKM=[UInt8](repeating: 0,count: 64)

    for i in 0..<64 {
        //IKM[i]=UInt8(i+1)
        IKM[i]=rng.getByte()
    }

    var res=bls48556.BLS256.initit()
    if res != 0 {
        print("Failed to Initialize\n")
        return
    }
    res=bls48556.BLS256.KeyPairGenerate(IKM,&S,&W)
    if res != 0 {
        print("Failed to Generate Keys\n")
        return
    }
    print("Testing bls48556 curve\n")      
    print("Private key : 0x",terminator: "");  printBinary(S);
    print("Public  key : 0x",terminator: "");  printBinary(W);
    let mess="This is a test message"
    print("Signing string - "+mess)
    
    bls48556.BLS256.core_sign(&SIG,[UInt8](mess.utf8),S)
    print("Signature : 0x",terminator: "");  printBinary(SIG);

    res=bls48556.BLS256.core_verify(SIG,[UInt8](mess.utf8),W)
    if res == 0 {
        print("Signature is OK\n");
    } else {
        print("Signature is *NOT* OK\n");
    }
}

print("Testing Boneh-Lynn-Shacham signature code\n");


var RAW=[UInt8](repeating: 0,count: 100)
var rng=RAND()
    
rng.clean();
for i in 0 ..< 100 {RAW[i]=UInt8(i&0xff)}
rng.seed(100,RAW)

TestBLS_bn254(&rng)
TestBLS_bls12383(&rng)
TestBLS_bls24479(&rng)
TestBLS_bls48556(&rng)
